lib/commands/push.ts

```diff
@@ -344,6 +305,143 @@ export default class PushCmd extends Command {
 		}
 	}
 
+	async pushToCloud(
+		options: FlagsDef,
+		sdk: BalenaSDK,
+		appOrDevice: string,
+		dockerfilePath: string,
+		registrySecrets: RegistrySecrets,
+		convertEol: boolean,
+		source: string,
+		nogitignore: boolean,
+	) {
+		const _ = await import('lodash');
+		const remote = await import('../utils/remote-build');
+
+		// Check for invalid options
+		const localOnlyOptions: Array<keyof FlagsDef> = [
+			'nolive',
+			'service',
+			'system',
+			'env',
+		];
+		this.checkInvalidOptions(
+			localOnlyOptions,
+			options,
+			'is only valid when pushing to a local mode device',
+		);
+
+		const releaseTags = options['release-tag'] ?? [];
+		const releaseTagKeys = releaseTags.filter((_v, i) => i % 2 === 0);
+		const releaseTagValues = releaseTags.filter((_v, i) => i % 2 === 1);
+
+		releaseTagKeys.forEach((key) => {
+			if (key === '') {
+				throw new ExpectedError(`Error: Keys can't be empty "".`);
+			}
+			if (/\s/.test(key)) {
+				throw new ExpectedError(`Error: Keys can't have whitespaces in them.`);
+			}
+		});
+		if (releaseTagKeys.length !== releaseTagValues.length) {
+			releaseTagValues.push('');
+		}
+
+		const app = appOrDevice;
+		await Command.checkLoggedIn();
+		const [token, baseUrl, owner] = await Promise.all([
+			sdk.auth.getToken(),
+			sdk.settings.get('balenaUrl'),
+			this.getAppOwner(sdk, app),
+		]);
+
+		const opts = {
+			dockerfilePath,
+			emulated: options.emulated || false,
+			multiDockerignore: options['multi-dockerignore'] || false,
+			nocache: options.nocache || false,
+			registrySecrets,
+			headless: options.detached || false,
+			convertEol,
+		};
+		const args = {
+			app,
+			owner,
+			source,
+			auth: token,
+			baseUrl,
+			nogitignore,
+			sdk,
+			opts,
+		};
+		const releaseId = await remote.startRemoteBuild(args);
+		if (releaseId) {
+			// Above we have checked that releaseTagKeys and releaseTagValues are of the same size
+			await Promise.all(
+				(_.zip(releaseTagKeys, releaseTagValues) as Array<
+					[string, string]
+				>).map(async ([key, value]) => {
+					await sdk.models.release.tags.set(releaseId, key, value);
+				}),
+			);
+		} else if (releaseTagKeys.length > 0) {
+			throw new Error(stripIndent`
+				The remote build has started or concluded, but a release ID could not be parsed
+				out of the builder's output. As a result, the release tags have not been set.`);
```

Comment:
I had suggested this error message before we decided to disallow using `--detached` in combination with `--release-tags`. In this case, the build is necessarily "concluded" (except for bugs, incorrect error handling...) and the message could be simplified a bit:
```
			throw new Error(stripIndent`
				A release ID could not be parsed out of the builder's output.
				As a result, the release tags have not been set.`);
```